File: packages/core/src/db/tags.ts
import type Database from 'better-sqlite3';
import type { Tag, TagColor } from '@qlan-ro/mainframe-types';
import { validateTagName } from '../lib/validate-tag-name.js';
import { hashTagColor } from '../lib/tag-color.js';

export class TagsRepository {
  constructor(private db: Database.Database) {}

  private normalize(name: string): string {
    return name.trim().toLowerCase();
  }

  list(): Tag[] {
    const rows = this.db.prepare('SELECT name, color, created_at as createdAt FROM tags ORDER BY name').all() as Tag[];
    return rows;
  }

  get(name: string): Tag | null {
    const normalized = this.normalize(name);
    const row = this.db
      .prepare('SELECT name, color, created_at as createdAt FROM tags WHERE name = ?')
      .get(normalized) as Tag | undefined;
    return row ?? null;
  }

  /** Idempotent upsert. Returns the existing row if present, else creates with auto color. */
  upsert(rawName: string, color?: TagColor): Tag {
    const v = validateTagName(rawName);
    if (!v.ok) throw new Error(v.error);
    const existing = this.get(v.normalized);
    if (existing) return existing;
    const finalColor: TagColor = color ?? hashTagColor(v.normalized);
    const now = new Date().toISOString();
    this.db.prepare('INSERT INTO tags (name, color, created_at) VALUES (?, ?, ?)').run(v.normalized, finalColor, now);
    return { name: v.normalized, color: finalColor, createdAt: now };
  }

  setColor(name: string, color: TagColor): void {
    const normalized = this.normalize(name);
    const info = this.db.prepare('UPDATE tags SET color = ? WHERE name = ?').run(color, normalized);
    if (info.changes === 0) throw new Error(`Tag not found: ${normalized}`);
  }

  /** Atomic rename. If `to` already exists, merges associations and drops `from`. */
  rename(fromRaw: string, toRaw: string): void {
    const from = this.normalize(fromRaw);
    const v = validateTagName(toRaw);
    if (!v.ok) throw new Error(v.error);
    const to = v.normalized;
    if (from === to) return;
    const tx = this.db.transaction(() => {
      const target = this.get(to);
      if (target) {
        // Merge: redirect chat_tags then delete `from` registry row.
        this.db
          .prepare(
            'INSERT OR IGNORE INTO chat_tags (chat_id, tag, source, created_at) ' +
              'SELECT chat_id, ?, source, created_at FROM chat_tags WHERE tag = ?',
          )
          .run(to, from);
        this.db.prepare('DELETE FROM chat_tags WHERE tag = ?').run(from);
        this.db.prepare('DELETE FROM tags WHERE name = ?').run(from);
      } else {
        // Plain rename — ON UPDATE CASCADE moves chat_tags rows.
        this.db.prepare('UPDATE tags SET name = ? WHERE name = ?').run(to, from);
      }
    });
    tx();
  }

  remove(name: string): void {
    const normalized = this.normalize(name);
    const tx = this.db.transaction(() => {
      this.db.prepare('DELETE FROM chat_tags WHERE tag = ?').run(normalized);
      const info = this.db.prepare('DELETE FROM tags WHERE name = ?').run(normalized);
      if (info.changes === 0) throw new Error(`Tag not found: ${normalized}`);
    });
    tx();
  }
}

File: packages/core-rs/crates/mainframe-db/src/tags.rs
//! Ported from `packages/core/src/db/tags.ts`.

use std::rc::Rc;

use mainframe_runtime::time::now_iso8601;
use mainframe_types::tags::{Tag, TagColor};
use rusqlite::Connection;

use crate::DbError;
use crate::tag_color::hash_tag_color;
use crate::validate_tag_name::{ValidateResult, validate_tag_name};

pub struct TagsRepository {
    db: Rc<Connection>,
}

fn parse_tag_color(s: String) -> Result<TagColor, DbError> {
    Ok(serde_json::from_value(serde_json::Value::String(s))?)
}

impl TagsRepository {
    pub fn new(db: Rc<Connection>) -> Self {
        Self { db }
    }

    fn normalize(&self, name: &str) -> String {
        name.trim().to_lowercase()
    }

    pub fn list(&self) -> Result<Vec<Tag>, DbError> {
        let mut stmt = self
            .db
            .prepare("SELECT name, color, created_at as createdAt FROM tags ORDER BY name")?;
        let mut rows = stmt.query([])?;
        let mut out = Vec::new();
        while let Some(row) = rows.next()? {
            out.push(Tag {
                name: row.get("name")?,
                color: parse_tag_color(row.get("color")?)?,
                created_at: row.get("createdAt")?,
            });
        }
        Ok(out)
    }

    pub fn get(&self, name: &str) -> Result<Option<Tag>, DbError> {
        let normalized = self.normalize(name);
        let mut stmt = self
            .db
            .prepare("SELECT name, color, created_at as createdAt FROM tags WHERE name = ?")?;
        let mut rows = stmt.query([normalized])?;
        match rows.next()? {
            Some(row) => Ok(Some(Tag {
                name: row.get("name")?,
                color: parse_tag_color(row.get("color")?)?,
                created_at: row.get("createdAt")?,
            })),
            None => Ok(None),
        }
    }

    /// Idempotent upsert. Returns the existing row if present, else creates with auto color.
    pub fn upsert(&self, raw_name: &str, color: Option<TagColor>) -> Result<Tag, DbError> {
        let normalized = match validate_tag_name(raw_name) {
            ValidateResult::Ok { normalized } => normalized,
            ValidateResult::Err { error } => return Err(DbError::Message(error)),
        };
        if let Some(existing) = self.get(&normalized)? {
            return Ok(existing);
        }
        let final_color = color.unwrap_or_else(|| hash_tag_color(&normalized));
        let now = now_iso8601();
        self.db.execute(
            "INSERT INTO tags (name, color, created_at) VALUES (?, ?, ?)",
            rusqlite::params![normalized, crate::enum_to_db_string(&final_color)?, now],
        )?;
        Ok(Tag {
            name: normalized,
            color: final_color,
            created_at: now,
        })
    }

    pub fn set_color(&self, name: &str, color: TagColor) -> Result<(), DbError> {
        let normalized = self.normalize(name);
        let changes = self.db.execute(
            "UPDATE tags SET color = ? WHERE name = ?",
            rusqlite::params![crate::enum_to_db_string(&color)?, normalized],
        )?;
        if changes == 0 {
            return Err(DbError::Message(format!("Tag not found: {normalized}")));
        }
        Ok(())
    }

    /// Atomic rename. If `to` already exists, merges associations and drops `from`.
    pub fn rename(&self, from_raw: &str, to_raw: &str) -> Result<(), DbError> {
        let from = self.normalize(from_raw);
        let to = match validate_tag_name(to_raw) {
            ValidateResult::Ok { normalized } => normalized,
            ValidateResult::Err { error } => return Err(DbError::Message(error)),
        };
        if from == to {
            return Ok(());
        }
        let tx = self.db.unchecked_transaction()?;
        if self.get(&to)?.is_some() {
            // Merge: redirect chat_tags then delete `from` registry row.
            tx.execute(
                "INSERT OR IGNORE INTO chat_tags (chat_id, tag, source, created_at) \
                 SELECT chat_id, ?, source, created_at FROM chat_tags WHERE tag = ?",
                rusqlite::params![to, from],
            )?;
            tx.execute("DELETE FROM chat_tags WHERE tag = ?", [&from])?;
            tx.execute("DELETE FROM tags WHERE name = ?", [&from])?;
        } else {
            // Plain rename — ON UPDATE CASCADE moves chat_tags rows.
            tx.execute(
                "UPDATE tags SET name = ? WHERE name = ?",
                rusqlite::params![to, from],
            )?;
        }
        tx.commit()?;
        Ok(())
    }

    pub fn remove(&self, name: &str) -> Result<(), DbError> {
        let normalized = self.normalize(name);
        let tx = self.db.unchecked_transaction()?;
        tx.execute("DELETE FROM chat_tags WHERE tag = ?", [&normalized])?;
        let changes = tx.execute("DELETE FROM tags WHERE name = ?", [&normalized])?;
        if changes == 0 {
            return Err(DbError::Message(format!("Tag not found: {normalized}")));
        }
        tx.commit()?;
        Ok(())
    }
}

// PORT STATUS: src/db/tags.ts (80 lines)
// confidence: high
// notes: validate_tag_name / hash_tag_color are imported from the sibling
// modules relocated into this crate (§2.15). `throw new Error(v.error)` /
// "Tag not found: X" become DbError::Message with byte-identical strings (the
// ported tests assert /reserved/i and /not found/i). rename()/remove() use
// unchecked_transaction() (Rc<Connection>); on the "Tag not found" early return
// the tx drops → ROLLBACK, matching the TS transaction that never commits. Tag
// color is stored as its serde string and parsed back via serde. Tests in
// tests/tags.rs.
// todos: 0
